File: extension/src/hal-drive.ts
// Minimal Drive REST calls needed for clip sync — upload a blob, fetch one
// back. Scope is drive.file (see hal-auth.ts), so this only ever touches
// files ctrl+freak itself created.

const HAL_DRIVE_UPLOAD_URL =
  "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart";
const HAL_DRIVE_FILES_URL = "https://www.googleapis.com/drive/v3/files";
const HAL_DRIVE_FOLDER_NAME = "Ctrl+Freak";
const HAL_DRIVE_FOLDER_MIME = "application/vnd.google-apps.folder";

// Cached per popup/tab lifetime — cheap enough to refetch if it's ever wrong,
// no need to persist across reloads.
let halFolderIdCache: string | null = null;

// Everything used to land loose in the root of My Drive with no way to find
// it again. Every upload now goes into one "Ctrl+Freak" folder instead.
async function halGetOrCreateAppFolder(accessToken: string): Promise<string> {
  if (halFolderIdCache) return halFolderIdCache;

  const query = encodeURIComponent(
    `name = '${HAL_DRIVE_FOLDER_NAME}' and mimeType = '${HAL_DRIVE_FOLDER_MIME}' and trashed = false`,
  );
  const searchResponse = await fetch(
    `${HAL_DRIVE_FILES_URL}?q=${query}&spaces=drive&fields=files(id)`,
    { headers: { Authorization: `Bearer ${accessToken}` } },
  );
  if (searchResponse.ok) {
    const data = (await searchResponse.json()) as { files: { id: string }[] };
    if (data.files.length > 0) {
      halFolderIdCache = data.files[0].id;
      return halFolderIdCache;
    }
  }

  const createResponse = await fetch(HAL_DRIVE_FILES_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${accessToken}`, "Content-Type": "application/json" },
    body: JSON.stringify({ name: HAL_DRIVE_FOLDER_NAME, mimeType: HAL_DRIVE_FOLDER_MIME }),
  });
  if (!createResponse.ok) {
    throw new Error(`Drive folder create failed: ${createResponse.status} ${await createResponse.text()}`);
  }
  const created = (await createResponse.json()) as { id: string };
  halFolderIdCache = created.id;
  return halFolderIdCache;
}

export async function halUploadFileToDrive(
  accessToken: string,
  blob: Blob,
  name: string,
): Promise<string> {
  const folderId = await halGetOrCreateAppFolder(accessToken);
  const boundary = "hal_drive_boundary_" + crypto.randomUUID();
  const metadata = JSON.stringify({ name, parents: [folderId] });

  const body = new Blob(
    [
      `--${boundary}\r\n`,
      "Content-Type: application/json; charset=UTF-8\r\n\r\n",
      metadata,
      `\r\n--${boundary}\r\n`,
      `Content-Type: ${blob.type || "application/octet-stream"}\r\n\r\n`,
      blob,
      `\r\n--${boundary}--`,
    ],
    { type: `multipart/related; boundary=${boundary}` },
  );

  const response = await fetch(HAL_DRIVE_UPLOAD_URL, {
    method: "POST",
    headers: { Authorization: `Bearer ${accessToken}` },
    body,
  });

  if (!response.ok) {
    throw new Error(`Drive upload failed: ${response.status} ${await response.text()}`);
  }

  const result = (await response.json()) as { id: string };
  return result.id;
}

export async function halFetchDriveFileBlob(
  accessToken: string,
  driveFileId: string,
): Promise<Blob> {
  const response = await fetch(
    `${HAL_DRIVE_FILES_URL}/${driveFileId}?alt=media`,
    { headers: { Authorization: `Bearer ${accessToken}` } },
  );
  if (!response.ok) {
    throw new Error(`Drive fetch failed: ${response.status} ${await response.text()}`);
  }
  return response.blob();
}

File: android/app/src/main/java/com/hal/ctrlfreak/drive/HalDrive.kt
package com.hal.ctrlfreak.drive

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException

// Mirrors extension/src/hal-drive.ts — same two REST calls, same
// drive.file scope (see HalAuth.kt), same multipart upload shape.

private val halHttpClient = OkHttpClient()

suspend fun halUploadFileToDrive(
    accessToken: String,
    bytes: ByteArray,
    mimeType: String,
    name: String,
): String = withContext(Dispatchers.IO) {
    val metadata = JSONObject().put("name", name).toString()
        .toRequestBody("application/json; charset=UTF-8".toMediaType())
    val body = MultipartBody.Builder()
        .setType(MultipartBody.FORM)
        .addPart(metadata)
        .addPart(bytes.toRequestBody(mimeType.toMediaType()))
        .build()

    val request = Request.Builder()
        .url("https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart")
        .header("Authorization", "Bearer $accessToken")
        .post(body)
        .build()

    halHttpClient.newCall(request).execute().use { response ->
        if (!response.isSuccessful) {
            throw IOException("Drive upload failed: ${response.code} ${response.body?.string()}")
        }
        JSONObject(response.body!!.string()).getString("id")
    }
}

suspend fun halFetchDriveFileBytes(accessToken: String, driveFileId: String): ByteArray =
    withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("https://www.googleapis.com/drive/v3/files/$driveFileId?alt=media")
            .header("Authorization", "Bearer $accessToken")
            .build()

        halHttpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Drive fetch failed: ${response.code} ${response.body?.string()}")
            }
            response.body!!.bytes()
        }
    }
